### ISSUE TITLE
gprecoverseg pg_rewind runs too long

### ISSUE THREAD

**Issue by diarworld** [greenplum-db/gpdb#9466]
Hi, all!
We have experienced problem when we need to recover large amount of segments with gprecoverseg tool. gprecoverseg have -B key witch described as:
> The number of segments to recover in parallel. If not specified, the utility will start up to four parallel processes depending on how many segment instances it needs to recover.

This key work correctly when we run recovery with -F (full recovery) key.

But this option does not work with -r (rebalance segments) key, when we switch primaries to primary role. All pg_rewind proccesses runs consequentially for each failed segment:
```
20200120:14:21:39:006970 gprecoverseg:mdw01:gpadmin-[INFO]:-334 segment(s) to recover
20200120:14:21:39:006970 gprecoverseg:mdw01:gpadmin-[INFO]:-Ensuring 334 failed segment(s) are stopped
20200120:14:24:46:006970 gprecoverseg:mdw01:gpadmin-[INFO]:-Ensuring that shared memory is cleaned up for stopped segments
20200120:14:24:46:006970 gprecoverseg:mdw01:gpadmin-[INFO]:-Updating configuration with new mirrors
20200120:14:24:48:006970 gprecoverseg:mdw01:gpadmin-[INFO]:-Updating mirrors
20200120:14:24:48:006970 gprecoverseg:mdw01:gpadmin-[INFO]:-Running pg_rewind on required mirrors
```

```
[gpadmin@mdw01 ~]$ gpssh -f segments -v -e 'ps aux | grep [r]ewind'
...
[sdw12] ps aux | grep [r]ewind
[sdw14] ps aux | grep [r]ewind
[sdw08] ps aux | grep [r]ewind
[sdw05] ps aux | grep [r]ewind
[sdw09] ps aux | grep [r]ewind
[sdw13] ps aux | grep [r]ewind
[sdw07] ps aux | grep [r]ewind
[sdw18] ps aux | grep [r]ewind
[sdw06] ps aux | grep [r]ewind
[sdw04] ps aux | grep [r]ewind
[sdw02] ps aux | grep [r]ewind
[sdw17] ps aux | grep [r]ewind
[sdw16] ps aux | grep [r]ewind
[sdw03] ps aux | grep [r]ewind
[sdw03] gpadmin  10640  0.0  0.0 113176  1592 ?        Ss   16:22   0:00 bash -c . /usr/local/greenplum-db/./greenplum_path.sh; [ -f /data1/mirror/gpseg30/recovery.conf ] || PGOPTIONS="-c gp_session_role=utility" /usr/local/greenplum-db/./bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="host=sdw02 port=10012 dbname=template1 application_name=__gprecoverseg_pg_rewind__" --target-pgdata=/data1/mirror/gpseg30 2>&1
[sdw03] gpadmin  10657  0.5  0.0 201036 27800 ?        S    16:22   0:00 /usr/local/greenplum-db/./bin/pg_rewind --write-recovery-conf --slot=internal_wal_replication_slot --source-server=host=sdw02 port=10012 dbname=template1 application_name=__gprecoverseg_pg_rewind__ --target-pgdata=/data1/mirror/gpseg30
[sdw01] ps aux | grep [r]ewind
[sdw10] ps aux | grep [r]ewind
[sdw20] ps aux | grep [r]ewind
[sdw22] ps aux | grep [r]ewind
[sdw21] ps aux | grep [r]ewind
[sdw19] ps aux | grep [r]ewind
[sdw11] ps aux | grep [r]ewind
[sdw15] ps aux | grep [r]ewind
[INFO] completed successfully
```

Recovery of one segment required up to 15 mins, so full incremental recovery will take more than 60 hours.

Could run pg_rewind on each segment in parralel when we use -B key for recovery?

### Greenplum version or build

Greenplum Database 6.0.1 OSS build

### OS version and uname -a

CentOS 7 kernel 3.10.0-957.12.1.el7.x86_64

### Installation information ( pg_config )

BINDIR = /usr/local/greenplum-db-6.0.1/bin
DOCDIR = /usr/local/greenplum-db-6.0.1/share/doc/postgresql
HTMLDIR = /usr/local/greenplum-db-6.0.1/share/doc/postgresql
INCLUDEDIR = /usr/local/greenplum-db-6.0.1/include
PKGINCLUDEDIR = /usr/local/greenplum-db-6.0.1/include/postgresql
INCLUDEDIR-SERVER = /usr/local/greenplum-db-6.0.1/include/postgresql/server
LIBDIR = /usr/local/greenplum-db-6.0.1/lib
PKGLIBDIR = /usr/local/greenplum-db-6.0.1/lib/postgresql
LOCALEDIR = /usr/local/greenplum-db-6.0.1/share/locale
MANDIR = /usr/local/greenplum-db-6.0.1/man
SHAREDIR = /usr/local/greenplum-db-6.0.1/share/postgresql
SYSCONFDIR = /usr/local/greenplum-db-6.0.1/etc/postgresql
PGXS = /usr/local/greenplum-db-6.0.1/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--disable-debug-extensions' '--disable-tap-tests' '--enable-orca' '--with-zstd' '--with-gssapi' '--with-libxml' '--with-perl' '--with-python' '--with-openssl' '--with-pam' '--with-ldap' '--enable-gpperfmon' '--prefix=/usr/local/greenplum-db-oss' '--mandir=/usr/local/greenplum-db-oss/man' 'CC=gcc' 'CFLAGS=-O3 -fargument-noalias-global -fno-omit-frame-pointer -g'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2 -I/usr/local/greenplum-db-6.0.1/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -O3 -fargument-noalias-global -fno-omit-frame-pointer -g -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/usr/local/greenplum-db-6.0.1/include
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/usr/local/greenplum-db-oss/lib',--enable-new-dtags -L/usr/local/greenplum-db-6.0.1/lib
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lxml2 -lpam -lrt -lgssapi_krb5 -lzstd -lrt -lcrypt -ldl -lm -L/usr/local/greenplum-db-6.0.1/lib
VERSION = PostgreSQL 9.4.24

### Expected behavior

Parallel incremental recovery.

### Actual behavior

Consequential recovery.

### Step to reproduce the behavior

`gprecoverseg -r -B 64` when failed segments count is large (in our case 300+ :) ).

**Comment by pivotal-ning-yu** [comment]
@greenplum-db/cluster-management could you check if could we let `gprecoverseg -r` run concurrently, please?

**Comment by Eulerizeit** [comment]
@diarworld have you looked at the code? Do you have a patch to suggest a fix.

**Comment by asimrp** [comment]
@diarworld FYI, I am working on a fix this.  Work in progress: https://github.com/asimrp/gpdb/commit/90f7300645e28301ebdaa40e1fda4a37fcb05c3f

**Comment by asimrp** [comment]
@diarworld this issue is fixed in 6X_STABLE: 3ddbed8eccb27dddd397848a800778c16cbcdc90  It should be available in the next release or you can try building the latest 6X_STABLE source to try it out.
